### ISSUE TITLE
Basic `async` support

### ISSUE THREAD

**Issue by septatrix** [pallets/click#2033]
This is a resumption of #85 which was closed more than _7 years ago_ when `asyncio` was at a _completely_ different stage in the python ecosystem. The goal of this issue is basically to enable users to do the following:

```python3
@click.command()
async def cli():
    await asyncio.sleep(1)
    click.echo("Hello World")

if __name__ == "__main__":
    cli()
```

Currently a workaround is to wrap every such function with a decorator like the following:

```python3
def make_sync(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        return asyncio.run(func(*args, **kwargs))
    return wrapper

@click.command()
@make_sync
async def cli():
    ...
```

The unclean aspect of the decorator approach is that it requires additional code which could be redundant  as click is able to simply infer via introspection that a function is a coroutine and handle accordingly.
This issue is only about the bare minimum to get such simple commands working. What this issue is not about is async-ifying most of the API like it is done in [`asyncclick`](https://github.com/python-trio/asyncclick) (a fork which also makes e.g. ctx and autocompletion async).

Arguments which have been brought up in the past by @mitsuhiko are that (1) this does not belong in click, (2) one would need to get access to the loop anyhow for complex use cases and (3) that there would inevitably be some people demanding support for twisted, trio etc. I would like to address each of them:

1. I see no reason why this does not belong in click. `async`/`await` is a first-class language feature in python and should be supported where feasible. Many CLI tools perform I/O, be it networking, file access or spawning subprocesses. All of these things can benefit from coroutines.
2. This is not really an argument as one can still get the current loop using `asyncio.get_event_loop()`/`asyncio.get_running_loop()` - previous cases where the loop itself may have been passed as an argument are [long made redundant](https://stackoverflow.com/questions/60312374/what-are-all-these-deprecated-loop-parameters-in-asyncio).
3. Those people can still use `asyncclick` which supports different event loops. However depending on the implementation it could be made possible to allow subclassing and overwriting some methods for increased compatibility.

Another reason why one might not want this might be decreased maintainability. However as the simplest implementations is 3 lines long I am not sure if this counts. Python 3.6 support would add another 2 lines.
On the other hand there are confused users when they get exceptions like `sys:1: RuntimeWarning: coroutine 'main' was never awaited` due to click lacking async support and users who need to get creative when the above decorator approach does not work anymore (when using some more involved type hints).

Other pallets projects which turned down async support years ago now support it like [Flask (since 2.0)](https://flask.palletsprojects.com/en/2.0.x/async-await/) and [Jinja (since 2.9)](https://jinja.palletsprojects.com/en/3.0.x/api/#async-support). They too only implement a minimum to work with it but that is sufficient for most people.

**Comment by mitsuhiko** [comment]
I agree with this generally. The ecosystem is in a vastly different space now than it was back then. That said, native support is a bit tricker because click allows to invoke things recursively and it's unclear if the expectation is that async -> sync is happening at any level or toplevel only.

The wrapping that is currently recommended effectively makes the async stop early and I'm not sure this is correct to do for a first class version.

**Comment by makkus** [comment]
> That said, native support is a bit tricker because click allows to invoke things recursively and it's unclear if the expectation is that async -> sync is happening at any level or toplevel only.

I agree that this is going to be tricky to get right. Personally, if this was implemented, I would expect async to be available at every stage throughout click. Even in the 'assemble sub-commands' stage, when calculating completions, when doing validation, etc.

I was using asyncclick for a project before, and in general it worked very well (I esp. liked that it used anyio under the hood to support multiple event loops). The one problem that was tricky to get implemented was that I wanted to dynamically discover sub-commands according to some json files on the filesystem, and I wanted to do all of this async. I kinda got it working, but don't ask me to show you the code... If I did that again, I think I'd just use a thread-pool, 'normal' click, and be done with it.

**Comment by davidism** [comment]
Supporting "everything" runs into the same problem as Flask: everything about Click's API is sync, so we'd have to essentially duplicate every object to have async versions of all their methods. And it's usually expected that both sync and async can be intermixed, which runs into the problem that we need to at *any* point in Click's dynamic resolution be able to switch execution contexts, or choose the right function based on the current execution context. Doing this sort of detection and switching also adds overhead. For a CLI app, responsiveness from the start is important, so this is not ideal.

For example, if type validation can be async, that must mean that a custom `ParamType` can define either or both of `def convert` and `async def async_convert`, then we need to decide which to run. Do we run `convert` because the command it's part of doesn't have an `async def` callback? Or do we always prefer starting an async loop if both are defined? What if a previous option in the processing pipeline has an `async def` callback, do we now prefer async because it already had to start a loop, even if the command is sync?

While I'm generally fine with the idea of supporting async, I really need to understand what the common use case we should solve for is. What *specifically* do people want to do that they can't do right now? Not "type conversion" but "I have to do this specific type conversion that absolutely requires async to be useful."

**Comment by mitsuhiko** [comment]
> Supporting "everything" runs into the same problem as Flask: everything about Click's API is sync, so we'd have to essentially duplicate every object to have async versions of all their methods.

I think the problem here is not as bad as in Flask. There is only a sync entrypoint to the call chain and a "make async code silently block" behavior is more than possible since nobody expects click to be involved in async call chains. One could then imagine a secondary entrypoint which does not "upgrade" a call to sync or even force a call to be async at all times.

I'm kinda curious where the general opinion is in 2021 now about how async is working out in Python. Are we moving into a world where almost every single thing is turning into `async` or is a situation evolving where some things are clearly sync and others async?

**Comment by ThiefMaster** [comment]
I have to admit, my first thought was "why do you need async in click" because in the end parsing command line args isn't something that feels very async... sure, you might have some edge cases where you do I/O while parsing arguments, but in the end that can be done sync as well...

**Comment by mitsuhiko** [comment]
So the reason I stumbled upon this issues is that I had to invoke a completely unnecessary async call chain from a click function. But alas, apparently almost everything is turning into async now so I guess this war is lost :P

**Comment by makkus** [comment]
> I have to admit, my first thought was "why do you need async in click" because in the end parsing command line args isn't something that feels very async... sure, you might have some edge cases where you do I/O while parsing arguments, but in the end that can be done sync as well...

I can only speak for myself, but if I have an async code-base that needs a command-line interface, being able to use call my existing async functions directly would be great. Not for very simple cases, but as soon as I use the more advanced stuff that click allows for (like the dynamic sub-command creation), it would certainly make life easier.

I'm not arguing that click should support that use-case, because as others on this thread I suspect the effort (and possibly complexity) involved in supporting this comprehensively would be quite high, and therefor not a good trade-off to make. It might make more sense to create a new 'async-from-the-ground-up' cli toolkit, dunno.

But, if click supported this, I would definitely find a few good use-cases for it, I'm 100% certain!

As for only shipping a few utility methods, decorators: that'd probably be nice, but personally I don't think I'd use those. I'd either use asyncclick, or threads.

**Comment by septatrix** [comment]
I am not suggesting complete support for async throughout the whole codebase but only for the command functions themselves.
They could be wrapped similar to the currently recommended decorator. Such that users new to click who need to use an async command are not surprised by exceptions and do not need to look for the decorator workaround.
Again: Users who need async in all the other places (completion, arg parsing etc.) are probably better of with using the existing asyncclick. However for users who call some async functions do not have to spin up their own event loop for each such command.

This would make click <-> asyncclick similar to the relationship between Flask <-> Quart.

Alternatively to make this a bit more clear we could add a new `AsyncCommand` class (and corresponding `@async_command`) decorator which perform this wrapping explicitly instead of adding it implicitly to the current `Command` invocation method.

**Comment by chobeat** [comment]
I think this discussion is kinda stranded but I would like to add that it would be great if the solution could make `CLIRunner` run well with asyncio too. I've been using `click` with home-baked async commands for a while but I can't really test properly because CLIRunner and my loop (injected by pytest-asyncio) don't really go well together.

**Comment by davidism** [comment]
The most that we would be adding is having Click automatically start an event loop if a command callback is marked `async def`, so that `await` could be written directly in the command callback. There is nothing stopping you from using async library code in your commands now, you just need to do `value = asyncio.run(coroutine)` instead of `await coroutine()`.

**Comment by chobeat** [comment]
that's already what I'm doing but then it conflicts with pytest-asyncio because CLIRunner does weird stuff and the coroutine is never awaited

**Comment by davidism** [comment]
> CLIRunner does weird stuff and the coroutine is never awaited

I can't reproduce that issue. `asyncio.run` is what awaits coroutines in that case, not the Click command or runner.

**Comment by lewoudar** [comment]
Hi everyone, I don't read all the conversation but if you want to add async support, don't forget that asyncio is not the only coroutine library, I'm thinking about [trio](https://trio.readthedocs.io/en/stable/). You can also take inspiration from the [asyncclick](https://github.com/python-trio/asyncclick) project.

**Comment by makkus** [comment]
@lewoudar asyncclick and trio are mentioned multiple times in this thread.

**Comment by lewoudar** [comment]
@makkus yeah, sorry, but I was afraid when I look at his first example dealing with asyncio

**Comment by davidism** [comment]
If something were to get in, it would presumably work similar to Flask, where there is a method that handles spawning the event loop for coroutines, and can be overridden to use something besides asyncio.

**Comment by lewoudar** [comment]
Thanks for the clarification @davidism  :)

**Comment by romanzdk** [comment]
Is there any plan to implement async support? :)

**Comment by bckohan** [comment]
Like @pocek I'd like to be able to run click command chains asynchronously in a single process.

As far as I can tell, this is the only use case where native support in click is necessary. For all the other ones its just syntactic sugar.

**Comment by pocek** [comment]
> Like [@pocek](https://github.com/pocek) I'd like to be able to run click command chains asynchronously in a single process.
> 
> As far as I can tell, this is the only use case where native support in click is necessary. For all the other ones its just syntactic sugar.

Unfortunately there are many async-unaware constructs in click. Running async commands is just a tip of an iceberg.

- anything using `click.globals` is more or less broken, and this means a lot of things. I've tried to replace `threading.local` with `contextvars`, but failed miserably (my conclusion at the time was that it's not possible without a much broader rework)
- `Context.with_resource`, `Context.call_on_close` etc. can't handle async
- `Context` is a synchronous context manager and because click makes heavy use of this feature pretty much everywhere, it makes many things impossible
- probably a lot more.

By _handle async_ I mean work in an existing event loop, not just call `loop.run_until_complete()` (this is not a solution).

**Comment by makukha** [comment]
What if a conservative approach can be used at the moment? Adding a simple decorator like `@click.sync` won't require integration and much support effort, but will cover the most common use case:
```python
@click.command()
@click.sync
async def cmd():
    pass
```

with trivial implementation
```python
import asyncio
from functools import wraps

def sync(func):
    '''Decorator that wraps coroutine with asyncio.run.'''
    @wraps(f)
    def wrapper(*args, **kwargs):
        return asyncio.run(f(*args, **kwargs))
    return wrapper
```

I mean, just adding the snippet from OP to `click` module will already help a lot.

**Comment by winny-** [comment]
It sounds like perfection isn't going to ship.  In light of that, maybe we could have the proposed decorator + a little blurb in the click docs.  This might solve 95% of users' challenges using asyncio, without prowling github issues for ad-hoc solutions.
